src/viper/pretty.ml

```diff
@@ -128,4 +131,44 @@ let prog p =
     let ppf = Format.formatter_of_buffer b in
     Format.fprintf ppf "@[%a@]" pp_prog p;
     Format.pp_print_flush ppf ();
+    let marks = ref (List.rev_map (fun loc -> loc, loc) !marks, [], []) in
+    let pos = ref 0 in
+    let push line column = match !marks with
+        | (mot, vip) :: clos, ope, don -> marks := clos, (mot, { vip with left = { vip.left with line; column } }) :: ope, don
+        | _ -> assert false in
+    let pop line column = match !marks with
+        | clos, (mot, vip) :: ope, don -> marks := clos, ope, (mot, { vip with right = { vip.right with line; column } }) :: don
+        | _ -> assert false in
+    let line = ref 1 in
+    let examine = function
+    | '\n' -> line := !line + 1; pos := 0; '\n';
+    | '\017' -> push !line !pos; '\017'
+    | '\019' -> pop !line !pos; '\017'
+    | a -> pos := !pos + 1; a in
+    let clean = function
+    | '\017' -> false
+    | _ -> true in
+    let b = Buffer.(of_seq Seq.(filter clean (map examine (to_seq b)))) in
+    Printf.eprintf "\nLINES: %d\n" !line;
+    let dump = List.iter (fun (mot, vip) -> Printf.eprintf "(MOT: %d:%d...%d:%d) -> (VIP: %d:%d...%d:%d)\n" mot.left.line mot.left.column mot.right.line mot.right.column vip.left.line vip.left.column vip.right.line vip.right.column) in
+    let _, _, mapping = !marks in
+    dump mapping;
+    let inside { left; right } other =
+        left.file = other.left.file &&
+        right.file = other.right.file &&
+        (other.left.line, other.left.column) <= (left.line, left.column) &&
+        (right.line, right.column) <= (other.right.line, other.right.column) in
+    let lookup (r : Source.region) =
+        let tighten prev (mot, vip) =
+            if inside r vip
+            then Some mot
```

Comment:
How do you know this is innermost loc enclosing r? Because the mapping is ordered from wide to narrow ?

Reply:
Exactly, enclosing spans get closed later, so they end up earlier in the list (i.e. pushed to the front later).